pupil_src/shared_modules/marker_tracker_3d/camera_localizer.py

```diff
@@ -0,0 +1,40 @@
+from marker_tracker_3d import math
+from marker_tracker_3d.localization import Localization
+
+
+class CameraLocalizer:
+    def __init__(self, storage):
+        self.storage = storage
+
+        self.localization = Localization(
+            self.storage.camera_model, self.storage.marker_model
+        )
+        self.min_number_of_markers_per_frame_for_loc = 2
+
+    def update(self):
+        self.update_camera_extrinsics()
+        self.update_camera_pose_matrix()
+        self.update_camera_trace()
+
+    def update_camera_extrinsics(self):
+        if len(self.storage.markers) >= self.min_number_of_markers_per_frame_for_loc:
+            self.storage.camera_extrinsics = self.localization.get_camera_extrinsics(
+                self.storage.markers,
+                self.storage.marker_extrinsics,
+                self.storage.camera_extrinsics_previous,
+            )
+        else:
+            self.storage.camera_extrinsics = None
+
+        if self.storage.camera_extrinsics is not None:
+            # Do not set camera_extrinsics_previous to None to ensure a decent initial guess for the next solve_pnp call
+            self.storage.camera_extrinsics_previous = self.storage.camera_extrinsics
+
+    def update_camera_pose_matrix(self):
+        self.storage.camera_pose_matrix = math.get_camera_pose_mat(
+            self.storage.camera_extrinsics
+        )
+
+    def update_camera_trace(self):
```

Comment:
Does it make sense for this function to be included in this class? Adding the localization to the trace of localizations goes kinda beyond the logic of localizing. Maybe `storage.camera_extrinsics` could become a property in `storage` with a setter that also sets `storage.camera_extrinsics_previous` and adds to `storage.camera_trace`. Than the `CameraLocalizer` would only do the localization and the `storage` class would handle the logic adding this information to the state accordingly.

Reply:
If this logic moves into `storage` I think `CameraLocalizer` and `Localization` should be merged.

Reply:
> `storage.camera_extrinsics` could become a property in `storage` with a setter that also sets `storage.camera_extrinsics_previous` and adds to `storage.camera_trace.`

I think is is a good idea. Just a bit concened whether it is proper that `storage` handles the logic.

Reply:
I would suggest not to merge `CameraLocalizer` and `Localization` since `Localization` is also used in `VisibilityGraphs` for deciding a keyframe.